File: leetcode/typescript/000977.squares-of-a-sorted-array.ts
function sortedSquares(nums: number[]): number[] {
  const ans: number[] = Array(nums.length).fill(0);
  let l = 0;
  let r = nums.length - 1;
  let i = nums.length - 1;
  while (i >= 0) {
    if (nums[r] > Math.abs(nums[l])) {
      ans[i] = nums[r] * nums[r];
      r--;
    } else {
      ans[i] = nums[l] * nums[l];
      l++;
    }
    i--;
  }
  return ans;
}
// Time complexity: O(n)
// Space complexity: O(1)

File: leetcode/python/000977.squares-of-a-sorted-array.py
class Solution:
    # Time complexity: O(n)
    # Space complexity: O(n)
    def sortedSquares(self, nums: list[int]) -> list[int]:
        l, r = 0, len(nums) - 1
        res = [0] * len(nums)
        for i in range(len(res) - 1, -1, -1):
            if abs(nums[l]) >= abs(nums[r]):
                n = nums[l]
                l += 1
            else:
                n = nums[r]
                r -= 1
            res[i] = n * n
        return res
